rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/BookKeeper.java

```diff
@@ -54,9 +74,57 @@ public BookKeeper(Configuration conf)
     }
 
     @Override
-    public List<Boolean> getCacheStatus(String remotePath, long fileLength, long lastModified, long startBlock, long endBlock)
+    public List<com.qubole.rubix.bookkeeper.Location> getCacheStatus(String remotePath, long fileLength, long lastModified, long startBlock, long endBlock, int clusterType)
             throws TException
     {
+        if (clusterManager == null || currentNodeIndex == -1) {
+            String nodename = null;
+            synchronized (lock) {
+                try {
+                    nodename = InetAddress.getLocalHost().getCanonicalHostName();
+                }
+                catch (UnknownHostException e) {
+                    e.printStackTrace();
+                    log.warn("Could not get nodename", e);
+                }
+
+                if (clusterType == HADOOP2_CLUSTER_MANAGER.ordinal()) {
+                    clusterManager = new Hadoop2ClusterManager();
+                    clusterManager.initialize(conf);
+                    nodes = clusterManager.getNodes();
+                    splitSize = clusterManager.getSplitSize();
+                }
+                else if (clusterType == TEST_CLUSTER_MANAGER.ordinal()) {
+                    nodes = new ArrayList<>();
+                    nodes.add(nodename);
+                    splitSize = 64 * 1024 * 1024;
+                }
+
+                if (nodes != null) {
+                    nodeListSize = nodes.size();
+                    currentNodeIndex = nodes.indexOf(nodename);
```

Comment:
we swallowed the expection in nodename above, it will throw error here now if nodename is null.
Catch the case where nodename is null and in that case put the logic such that we fallback to DirectReads